var/spack/repos/builtin/packages/python/package.py

```diff
@@ -326,7 +326,7 @@ class Python(Package):
     # An in-source build with --enable-optimizations fails for python@3.X
     build_directory = "spack-build"
 
-    executables = [r"^python[\d.]*[mw]?$"]
+    executables = [r"^python\d?$"]
```

Comment:
Starting with Python 3.8, `pythonXm` is gone[^1]. And Python 3.7 is EOL, so I don't think we need to support detecting it, but maybe @tgamblin can correct me if I'm wrong.

I'm less certain about `pythonXw`. I believe there used to be a `pythonXd` too[^2]. I don't know if there's any situation in which `pythonX[dmw]` would exist but `pythonX` wouldn't.

[^1]: https://docs.python.org/3/whatsnew/3.8.html#build-and-c-api-changes
[^2]: https://stackoverflow.com/q/16675865/5828163

Reply:
It looks like this was first added by you in https://github.com/spack/spack/commit/9ab8521a1c09558339900041647b2466c2709de9, which is to say: no one went out of their way at some point to extend detection to e.g. `python3.7m`, it was just provided from the start (and I doubt there will be any complaints about constraining this).

In https://github.com/spack/spack/issues/40625#issuecomment-1771740699, you bring up that we prefer `pythonx.y` over `pythonx`, so Spack itself should not be getting confused about this; that makes me concerned that this may not fix https://github.com/spack/spack/issues/40625 (if it is indeed the build system looking through `/usr/` for Python).

@mwkrentel can you confirm that removing the extra entries from `packages.yaml` fixes #40625 for you?

Reply:
Wait thinking about this again, this will work because we'll just add `python3` (and will associate the correct version with it), vs. potentially adding an "other" `python3.x` that isn't associated with the `python3` executable.

In the case of #40625, I assume this will fix it by forcing Spack to build a suitable Python version.

Reply:
I think the fundamental assumption in this PR is that any python without a `python3` is incomplete if not broken, because *shebangs are written for `#!/usr/bin/env python3`*. So we will only *automatically* detect something that has a `python3`. For anything else, YMMV.

`executables` is only used for *detection* so it doesn’t have to care about the `m` suffix. `command` can try to care, to do a best-effort attempt at using any Python installation *without* `python3`, but for detection we do not care.

FWIW, IMO the expression here should just say `^python3$`, because something with *just* `python` is going to break things that have `python3` shebangs, and any other match for `^python\d$` is something weird and broken. There is no `python1` or `python4`. We don’t support `python2`…. 5 is right out, and I dunno wtf `python0` is but I don’t want to bother running it.

So in the spirit of, well, Python, I say 3 shall be the number thou shalt count and the number thou shalt count shall be 3.